File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean flag = false;
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {

    if (root == null) {
        return false;
    }

    if (root.val == subRoot.val) {
        if (dfs(root, subRoot)) {
            flag = true;
        }
    }

    isSubtree(root.left, subRoot);
    isSubtree(root.right, subRoot);

    return flag;
}
    
    public boolean dfs(TreeNode root, TreeNode subRoot){
        if(root==null && subRoot==null){
            return true;
        }
        else if(root!=null && subRoot==null){
            return false;
        }
        else if(root==null && subRoot!=null){
            return false;
        }
        else{
            if(root.val!=subRoot.val){
                return false;
            }
        }
        return dfs(root.left, subRoot.left) && dfs(root.right, subRoot.right);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
class Solution:

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:

        self.lca = TreeNode(0)

        def LCA(root: TreeNode, p: TreeNode, q: TreeNode) -> bool:
            if not root:
                return False

            left = LCA(root.left, p, q)
            right = LCA(root.right, p, q)

            if left and right:
                self.lca.val = root.val

            if root.val == p.val and (left or right):
                self.lca.val = root.val

            if root.val == q.val and (left or right):
                self.lca.val = root.val

            if root.val == p.val or (left or right):
                return True

            if root.val == q.val or (left or right):
                return True
            return left or right

        LCA(root, p, q)
        return self.lca